examples/pytorch/aml.py

```diff
@@ -0,0 +1,162 @@
+4. For multi-node, this script needs to be run in azure ml terminal not local.
+"""
+from time import sleep
+
+import numpy as np
+import numpy.testing as npt
+import ray
+from ray.train.torch import TorchTrainer
+from ray.air import session
+from ray.air.config import ScalingConfig, RunConfig
+
+from deepgnn.graph_engine.data.cora import CoraFull
+from deepgnn.graph_engine.snark.distributed import Server, Client as DistributedClient
+
+
+@ray.remote
+class ServerLock:
+    """Lock for server."""
+
+    def __init__(self, size):
+        """Initialize server lock."""
+        self.cache = {}
+        self.locks = [0 for _ in range(size)]
+
+    def set_lock(self, i):
+        """Set lock i."""
+        self.locks[i] = 1
+
+    def release_lock(self, i):
+        """Release lock i."""
+        self.locks[i] = 2
+
+    def wait_set(self):
+        """See if all locks set."""
+        return all([v == 1 for v in self.locks])
+
+    def wait_released(self):
+        """See if all locks released."""
+        return all([v == 2 for v in self.locks])
+
+    def put(self, x, y):
+        """Put value y at x."""
+        self.cache[x] = y
+
+    def get(self, x):
+        """Get value x."""
+        return self.cache.get(x)
+
+
+class ServerContext:
+    """Server context."""
+
+    def __init__(self, world_size: int, rank: int, address: str):
+        """Initialize server context."""
+        self.world_size = world_size
+        self.rank = rank
+        self.address = address
+        self.server_lock = server_lock
+
+    def __enter__(self):
+        """Enter server context."""
+        self.server_lock.put.remote(self.rank, self.address)
+        self.server_lock.set_lock.remote(self.rank)
+        while not ray.get(self.server_lock.wait_set.remote()):
+            sleep(1)
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """Exit server context."""
+        self.server_lock.release_lock.remote(self.rank)
+        while not ray.get(self.server_lock.wait_released.remote()):
+            sleep(1)
+
+    def get_address(self):
+        """Return addresses of all servers connected."""
+        return ray.get([self.server_lock.get.remote(i) for i in range(self.world_size)])
+
+
+def train_func(config: dict):
+    """Training loop for ray trainer."""
+    cora = CoraFull(num_partitions=session.get_world_size())
+
+    hostname = (
+        "127.0.0.1"
+        if config["is_unit_test"]
+        else ray._private.services.get_node_ip_address()
+    )
+    address = f"{hostname}:999{session.get_world_rank()}"
+
+    Server(address, cora.data_dir(), session.get_world_size(), 1)
+
+    with ServerContext(
+        session.get_world_size(), session.get_world_rank(), address
+    ) as lock:
+        cl = DistributedClient(lock.get_address())
+
+        features = cl.node_features(np.array([0, 1]), np.array([[0, 1]]), np.float32)
+        npt.assert_equal(features, np.array([[0.0], [0.0]]))
```

Comment:
we backfill with 0s if a feature is not found too. Could you add some non-default values(probably with feature_id 1, instead of 0)?

Reply:
For some reason, all node features for index 0 and 1 have no features up to length 121, even for dtype int64

Reply:
other node_ids might work too..

Reply:
It looks like this is a bug somehow, cora.features works but cl.features doesn't ever